client-core/src/multi_sig/builder.rs

```diff
@@ -0,0 +1,277 @@
+        }
+
+        Ok(nonce)
+    }
+
+    /// Adds a nonce from a public key to session.
+    pub fn add_nonce(&mut self, public_key: &PublicKey, nonce: &PublicKey) -> Result<()> {
+        self.session.add_nonce(public_key, nonce.clone())?;
+
+        Ok(())
+    }
+
+    /// Returns partial signature of current signer. Add the partial signature
+    /// to the session if the current signer has no partial signature added
+    /// before.
+    ///
+    /// This function will fail if nonces from all co-signers are not received.
+    pub fn partial_signature(&mut self) -> Result<H256> {
+        let partial_signature = self.session.partial_signature()?;
+        let public_key = &self.session.public_key.clone();
+
+        if !self.session.has_partial_signature(public_key)? {
+            self.session
+                .add_partial_signature(public_key, partial_signature)?;
+        }
+
+        Ok(partial_signature)
+    }
+
+    /// Adds a partial signature from a public key to session
+    pub fn add_partial_signature(
+        &mut self,
+        public_key: &PublicKey,
+        partial_signature: H256,
+    ) -> Result<()> {
+        self.session
+            .add_partial_signature(public_key, partial_signature)?;
+
+        Ok(())
+    }
+
+    /// Returns final signature.
+    ///
+    /// This function will fail if partial signatures from all co-signers are not received.
+    pub fn signature(&self) -> Result<SchnorrSignature> {
+        self.session.signature()
+    }
+
+    /// Returns public keys of all signers in this session
+    pub fn public_keys(&self) -> Vec<PublicKey> {
+        self.session.public_keys()
+    }
+
+    /// Returns true if nonce commitment for given public key is already set,
+    /// false otherwise.
+    pub fn has_nonce_commitment(&self, public_key: &PublicKey) -> Result<bool> {
+        self.session.has_nonce_commitment(public_key)
+    }
+
+    /// Returns true if nonce for given public key is already set, false
+    /// otherwise.
+    pub fn has_nonce(&self, public_key: &PublicKey) -> Result<bool> {
+        self.session.has_nonce(public_key)
+    }
+
+    /// Returns true if partial signature for given public key is already set,
+    /// false otherwise.
+    pub fn has_partial_signature(&mut self, public_key: &PublicKey) -> Result<bool> {
+        self.session.has_partial_signature(public_key)
+    }
+
+    /// Returns incompleted MultiSig session in bytes
+    pub fn to_incomplete(&self) -> Vec<u8> {
+        self.session.encode()
+    }
+
+    /// Restore MultiSig session builder from encoded incompleted bytes
+    pub fn from_incomplete(bytes: Vec<u8>) -> Result<Self> {
+        let session = MultiSigSession::decode(&mut bytes.as_slice()).chain(|| {
+            (
+                ErrorKind::DeserializationError,
+                "Unable to deserialize MultiSig session",
+            )
+        })?;
+
+        Ok(MultiSigBuilder { session })
+    }
+}
```

Comment:
Is this needed? I don't see it used anywhere.
If it's needed, comments should mention a warning for potential foot guns:

- the underlying multi-party protocol should be aborted if there's a problem detected at any step -- the expected recovery is to start a fresh session
- if there is an API to persist/serialize incomplete sessions, the user should better know to be careful about ensuring the latest session state, otherwise there's a potential "footgun":
  - user starts a session -- incomplete session state 0
  - ...
  - user continues a session -- incomplete session state N
  - an attacker (somehow) manages to force the user to restart the session from some incomplete session state before N
  - e.g. an attacker observed the user's nonce commitment and nonce, tweaked its own nonce and force the user to restart from the state before seeing the attacker's nonce commitment -- a variant of https://medium.com/blockstream/insecure-shortcuts-in-musig-2ad0d38a97da

Reply:
Yes it is intended for the `cro-nodelib`, but the scenario is a concern. I think @foreseaz should be aware of too when designing the `cro-nodelib`.

And I will add some warnings to the comment.